File: C++/Medium/Accepted/shortest-path-with-alternating-colors_1.cpp
class Solution {
public:
    vector<int> shortestAlternatingPaths(int n, vector<vector<int>>& redEdges, vector<vector<int>>& blueEdges) {
        
        vector<vector<vector<int>>> graph(n, vector<vector<int>>(2, vector<int>()));
        for (auto& edge : redEdges) {
            graph[edge[0]][0].push_back(edge[1]);
        }
        for (auto& edge : blueEdges) {
            graph[edge[0]][1].push_back(edge[1]);
        }
        vector<vector<int>> dist(n, vector<int>(2, INT_MAX));
        dist[0][0] = dist[0][1] = 0;
        queue<pair<int, int>> q;
        q.push({0, 0});
        q.push({0, 1});
        while (!q.empty()) {
            int node = q.front().first;
            int color = q.front().second;
            q.pop();
            for (int nei : graph[node][color]) {
                if (dist[nei][color ^ 1] == INT_MAX) {
                    dist[nei][color ^ 1] = dist[node][color] + 1;
                    q.push({nei, color ^ 1});
                }
            }
        }
        vector<int> ans(n, -1);
        for (int i = 0; i < n; i++) {
            if (dist[i][0] != INT_MAX || dist[i][1] != INT_MAX) {
                ans[i] = min(dist[i][0], dist[i][1]);
            }
        }
        return ans;
    }
};

File: Rust/Medium/Accepted/shortest-path-with-alternating-colors_1.rs
impl Solution {
    pub fn shortest_alternating_paths(n: i32, red_edges: Vec<Vec<i32>>, blue_edges: Vec<Vec<i32>>) -> Vec<i32> {
        
        let mut red = vec![vec![]; n as usize];
        let mut blue = vec![vec![]; n as usize];
        for e in red_edges {
            red[e[0] as usize].push(e[1] as usize);
        }
        for e in blue_edges {
            blue[e[0] as usize].push(e[1] as usize);
        }
        let mut ans = vec![-1; n as usize];
        ans[0] = 0;
        let mut q = std::collections::VecDeque::new();
        let mut visited = vec![vec![false; n as usize]; 2];
        q.push_back((0, 0));
        q.push_back((0, 1));
        visited[0][0] = true;
        visited[1][0] = true;
        let mut step = 0;
        while !q.is_empty() {
            step += 1;
            let size = q.len();
            for _ in 0..size {
                let (i, c) = q.pop_front().unwrap();
                let next = if c == 0 { &blue[i] } else { &red[i] };
                for &j in next {
                    if ans[j] == -1 {
                        ans[j] = step;
                    }
                    if !visited[1 - c][j] {
                        q.push_back((j, 1 - c));
                        visited[1 - c][j] = true;
                    }
                }
            }
        }
        ans
    }
}